src/main/flight/mixer.c

```diff
@@ -346,14 +346,92 @@ static void applyFlipOverAfterCrashModeToMotors(void)
+            if (rcCommandThrottle < motorConfig()->digitalIdleOffsetValue / 10000.0f) {
+                mixerRuntime.rpmLimiterI *= 1.0f / (1.0f + (pidGetDT() * 10.0f)); //slowly ramp down i term instead of resetting to avoid throttle pulsing cheats
+            } else if (!motorsSaturated) {
+                //don't let I term wind up if motors are saturated. Otherwise, motors may stay at high throttle even after low throttle is commanded
+                mixerRuntime.rpmLimiterI += smoothedRPMError * mixerRuntime.rpmLimiterIGain;
+            }
+            pidOutput = rpm_limiterP + mixerRuntime.rpmLimiterI + rpmLimiterD;
+        } else {
+            throttle = throttle * mixerRuntime.rpmLimiterExpectedThrottleLimit;
+            mixerRuntime.rpmLimiterI += smoothedRPMError * mixerRuntime.rpmLimiterIGain;
+            mixerRuntime.rpmLimiterI = MAX(mixerRuntime.rpmLimiterI, 0.0f);
+            pidOutput = rpm_limiterP + mixerRuntime.rpmLimiterI + rpmLimiterD;
+            if (pidOutput > 0.05) {
+                mixerRuntime.rpmLimiterExpectedThrottleLimit = 0.9994 * mixerRuntime.rpmLimiterExpectedThrottleLimit;
+            }
+            if (pidOutput < -0.05 && rcCommand[THROTTLE] > 1950 && !motorsSaturated) {
+                mixerRuntime.rpmLimiterExpectedThrottleLimit = 1.0006f * mixerRuntime.rpmLimiterExpectedThrottleLimit;
+                mixerRuntime.rpmLimiterExpectedThrottleLimit = MAX(mixerRuntime.rpmLimiterExpectedThrottleLimit, 1.0f);
+            }
+            pidOutput = MAX(pidOutput, 0.0f);
+        }
+        if (mixerRuntime.rpmLimiterInit) {
+            if (mixerConfig()->rpm_limiter_rpm_linearization) {
+                throttle = constrainf(-pidOutput, 0.01f, 1.0f);
+            } else {
+                throttle = constrainf(throttle-pidOutput, 0.01f, 1.0f);
+            }
+        }
+        mixerRuntime.rpmLimiterInit = true;
+        mixerRuntime.prevAverageRPM = averageRPM;
+        mixerRuntime.rpmLimiterPreviousSmoothedRPM = averageRPMSmoothed;
+        mixerRuntime.rpmLimiterPreviousSmoothedRPMError = smoothedRPMError;
+        mixerRuntime.rpmLimiterPreviousRPMLimit = scaledRPMLimit;
+        DEBUG_SET(DEBUG_RPM_LIMITER, 0, averageRPM);
+        DEBUG_SET(DEBUG_RPM_LIMITER, 1, smoothedRPMError);
+        DEBUG_SET(DEBUG_RPM_LIMITER, 2, mixerRuntime.rpmLimiterI * 100.0f);
+        DEBUG_SET(DEBUG_RPM_LIMITER, 3, rpmLimiterD * 10000.0f);
+    }
+}
+#endif
+
 static void applyMixToMotors(float motorMix[MAX_SUPPORTED_MOTORS], motorMixer_t *activeMixer)
 {
     // Now add in the desired throttle, but keep in a range that doesn't clip adjusted
     // roll/pitch/yaw. This could move throttle down, but also up for those low throttle flips.
     for (int i = 0; i < mixerRuntime.motorCount; i++) {
         float motorOutput = motorOutputMixSign * motorMix[i] + throttle * activeMixer[i].throttle;
 #ifdef USE_THRUST_LINEARIZATION
+    if (!mixerConfig()->rpm_limiter) {
```

Comment:
You should look into a way of allowing for thrust linearization and this code. Thrust linearization should still be allowed as thrust linear is meant to linearize the pid controller, and not meant to mess around with the thrust that is produced. Since this RPM limiter simply messes around with throttle.

Reply:
Gotcha, will remove this check

Reply:
It needs a little more than this as currently the thrust linear code will also mess around with the throttle (and the thrust linear was recoded so that it doesn't remove the extra throttle it adds correctly :( ). But there should be a way to make this happen.